File: app/src/ipc/export.ts
import { invoke } from "@tauri-apps/api/core";

const cache = new Map<string, Promise<string>>();

/** Higher-resolution image (data URI) for print export, cached per path. */
export function getExportImage(path: string): Promise<string> {
  let p = cache.get(path);
  if (!p) {
    p = invoke<string>("get_export_image", { path });
    cache.set(path, p);
  }
  return p;
}

export interface ExportFile {
  name: string;
  b64: string; // raw base64, no data: prefix
}

/** Write rendered files to `dir` (created if needed). Returns the dir. */
export function writeExport(dir: string, files: ExportFile[]): Promise<string> {
  return invoke<string>("write_export", { dir, files });
}

File: app/src-tauri/src/export.rs
//! Export (§4.6) — write rendered spread files (JPG/PDF) to disk.

use std::fs;
use std::path::Path;

use base64::{engine::general_purpose, Engine};
use serde::Deserialize;

#[derive(Deserialize)]
pub struct ExportFile {
    name: String,
    /// raw base64 (no data: prefix)
    b64: String,
}

/// Create `dir` (and parents) and write each file. Returns the directory path.
#[tauri::command]
pub async fn write_export(dir: String, files: Vec<ExportFile>) -> Result<String, String> {
    let p = Path::new(&dir);
    fs::create_dir_all(p).map_err(|e| format!("Tạo thư mục lỗi: {e}"))?;
    for f in files {
        let bytes = general_purpose::STANDARD
            .decode(f.b64.trim())
            .map_err(|e| format!("{}: {e}", f.name))?;
        fs::write(p.join(&f.name), bytes).map_err(|e| format!("{}: {e}", f.name))?;
    }
    Ok(dir)
}
